File: rid-derive/src/templates/vec.dart
/// extension Rid_ExtOnPointer{vec_type} on {dart_ffi}.Pointer<{ffigen_bind}.{vec_type}> {
///   int get length => {rid_ffi}.{fn_len_ident}(this);
///   int operator [](int idx) {
///     final len = this.length;
///     if (!(0 <= idx && idx < len)) {
///       throw AssertionError("Out of range access [$idx] on List of length $len");
///     }
///     return {rid_ffi}.{fn_get_ident}(this, idx);
///   }
///   Rid_{vec_type}_Iterable iter() => Rid_{vec_type}_Iterable(this);
/// }
/// 
/// class Rid_{vec_type}_Iterator implements Iterator<{dart_item_type}> {
///   int _currentIdx = -1;
///   final {dart_ffi}.Pointer<ffigen_bind.{vec_type}> _vec;
///   final int _limit;
/// 
///   Rid_{vec_type}_Iterator(this._vec) : _limit = _vec.length - 1;
/// 
///   int get current => _vec[_currentIdx];
/// 
///   bool moveNext() {
///     if (_currentIdx >= _limit) return false;
///     _currentIdx++;
///     return true;
///   }
/// }
/// 
/// class Rid_{vec_type}_Iterable with {dart_collection}.IterableMixin<{dart_item_type}> {
///   final {dart_ffi}.Pointer<ffigen_bind.{vec_type}> _vec;
///   Rid_Vec_u8_Iterable(this._vec);
/// 
///   Iterator<{dart_item_type}> get iterator => Rid_{vec_type}_Iterator(this._vec);
/// }

File: rid-derive/src/templates/vec.rs
use rid_common::{DART_COLLECTION, DART_FFI, FFI_GEN_BIND, RID_FFI};

use crate::dart::DartType;

const TEMPLATE: &str = std::include_str!("./vec.dart");

pub(crate) struct ImplementVec {
    pub(crate) vec_type: String,
    pub(crate) dart_item_type: DartType,
    pub(crate) fn_len_ident: String,
    pub(crate) fn_get_ident: String,
}

pub(crate) fn render(vec: &ImplementVec) -> String {
    let iterated_item_type = if vec.dart_item_type.is_primitive() {
        vec.dart_item_type.to_string()
    } else {
        format!(
            "{dart_ffi}.Pointer<{ffigen_bind}.{dart_item_type}>",
            dart_ffi = DART_FFI,
            ffigen_bind = FFI_GEN_BIND,
            dart_item_type = &vec.dart_item_type
        )
    };
    TEMPLATE
        .replace("{vec_type}", &vec.vec_type)
        .replace("{dart_item_type}", &vec.dart_item_type.to_string())
        .replace("{iterated_item_type}", &iterated_item_type)
        .replace("{fn_len_ident}", &vec.fn_len_ident)
        .replace("{fn_get_ident}", &vec.fn_get_ident)
        .replace("{ffigen_bind}", FFI_GEN_BIND)
        .replace("{dart_ffi}", DART_FFI)
        .replace("{rid_ffi}", RID_FFI)
        .replace("{dart_collection}", DART_COLLECTION)
}
